File: ProblemSolve/23.cs
/************************
 * Merge K Sorted Lists *
 ************************/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

public class Solution {
    public ListNode MergeKLists(ListNode[] lists) {
        ListNode ans = new ListNode();
        ListNode pointer = ans;        
                
        PriorityQueue<ListNode, int> pq = new PriorityQueue<ListNode, int>();
        
        for(int i=0; i<lists.Length; ++i){                        
            if(lists[i] != null)
                pq.Enqueue(lists[i], lists[i].val);
        }
        
        while(pq.TryDequeue(out ListNode list, out int val)){
            pointer.next = new ListNode(val);
            pointer = pointer.next;
            list = list.next;
            
            if(list != null){
                pq.Enqueue(list, list.val);
            }
        }
        
        return ans.next;
    }
}

File: ProblemSolve/23.cpp
/************************
 * Merge K Sorted Lists *
 ************************/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    struct Compare{
        bool operator()(pair<int, ListNode*> a, pair<int, ListNode*> b){
            return a.first > b.first;
        }
    };
    
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        ListNode* ans = new ListNode();
        ListNode* pointer = ans;
        
        priority_queue<pair<int, ListNode*>, vector<pair<int, ListNode*>>, Compare> pq;
        
        for(int i=0; i<lists.size(); ++i){
            if(lists[i] != NULL){
                pq.push(make_pair(lists[i]->val, lists[i]));
            }
        }
        
        while(!pq.empty()){
            pair<int, ListNode*> minimumListNVal = pq.top();
            pq.pop();
            pointer->next = new ListNode(minimumListNVal.second -> val);
            pointer = pointer -> next;
            minimumListNVal.second = minimumListNVal.second->next;
            
            if(minimumListNVal.second != NULL){
                pq.push(make_pair(minimumListNVal.second -> val, minimumListNVal.second));
            }
        }
        
        return ans->next;
    }
};